compiler/ast/trees.nim

```diff
@@ -44,61 +44,64 @@ proc sameFloatIgnoreNan(a, b: BiggestFloat): bool {.inline.} =
   ## ignores NaN semantics, but ensures 0.0 == -0.0, see #13730
   cast[uint64](a) == cast[uint64](b) or a == b
 
-proc exprStructuralEquivalent*(a, b: PNode; strictSymEquality=false): bool =
-  if a == b:
-    result = true
-  elif (a != nil) and (b != nil) and (a.kind == b.kind):
-    case a.kind
-    of nkSym:
-      if strictSymEquality:
-        result = a.sym == b.sym
-      else:
-        # don't go nuts here: same symbol as string is enough:
-        result = a.sym.name.id == b.sym.name.id
-    of nkIdent: result = a.ident.id == b.ident.id
-    of nkIntLiterals: result = a.intVal == b.intVal
-    of nkFloatLiterals: result = sameFloatIgnoreNan(a.floatVal, b.floatVal)
-    of nkStrLiterals: result = a.strVal == b.strVal
-    of nkCommentStmt: result = a.comment == b.comment
-    of nkEmpty, nkNilLit, nkType: result = true
-    of nkError:
-      unreachable()
-    of nkWithSons:
-      if a.len == b.len:
-        for i in 0..<a.len:
-          if not exprStructuralEquivalent(a[i], b[i],
-                                          strictSymEquality): return
-        result = true
+template makeTreeEquivalenceProc*(
+  name, relaxedKindCheck, symCheck, floatCheck, typeCheck, commentCheck) {.dirty.} =
+  # Defines a tree equivalence checking procedure.
+  # This skeleton is shared between all recursive
+  # PNode equivalence checks in the compiler code base
+  # It might be possible to unify more of them with each other.
```

Comment:
```suggestion
  ## Defines a tree equivalence checking procedure.
  ## This skeleton is shared between all recursive
  ## `PNode` equivalence checks in the compiler code base
  ## It might be possible to unify more of them with each other.
```